### ISSUE TITLE
layer._extent_data inconsistent between image and points/shapes/tracks

### ISSUE THREAD

**Issue by kevinyamauchi** [napari/napari#1686]
## 🐛 Bug
The values for `_extent_data` are calculated differently for the image layer and points/shapes/tracks. In the image layer, the maximum value for the extent is the `maximum index + 1` (i.e., the result of `data.shape`) and in the points/shapes/tracks layer it is the maximum index (i.e., the results of `np.max(data)`. It appears that our Dims model assumes the maximum value of the extent is `maximum_index + 1` so the range of the slider is incorrectly calculated for the points/shapes/tracks layers (it is 1 step less).

Is `layer._extent_data` defined as `(min_index, max_index + 1 )`, where `min_index `and `max_index` are the layer data array indices?

## To Reproduce
1. create 3D data with indices from 0 through 4 for axis 0 for image, points, shapes, and tracks layers
2. view those data in napari in 2D mode (axis slider is the 0th axis). Note that the axis slider correctly goes from 0-4
![Screenshot from 2020-10-03 13-24-08](https://user-images.githubusercontent.com/1120672/94990371-cccc7600-057b-11eb-9330-1b10fb2a8200.png)

3. view the `layer._extent_data` property for each layer

```python
viewer.layers['image']._extent_data
Out[1]: 
array([[   0.,    0.,    0.],
       [   5., 1000., 1000.]])

viewer.layers['points']._extent_data
Out[2]: 
array([[   0,  236,    0],
       [   4,  636, 1000]])

viewer.layers['tracks']._extent_data
Out[3]: 
array([[   0,  236,    0],
       [   4,  636, 1000]])
viewer.layers['shapes']._extent_data
Out[4]: 
array([[ 0., 10., 10.],
       [ 4., 20., 20.]])
```
4. If we  delete the image layer, the axis slider now goes from 0 - 3.
![Screenshot from 2020-10-03 13-14-07](https://user-images.githubusercontent.com/1120672/94990229-ac4fec00-057a-11eb-8190-69778009730b.png)

<details>
<summary>code</summary>

```python

test_im = np.random.random((5, 1000, 1000))

tracks_data = np.asarray([
    [1, 0, 236, 0],
    [1, 1, 236, 100],
    [1, 2, 236, 200],
    [1, 3, 236, 500],
    [1, 4, 236, 1000],
    [2, 0, 436, 0],
    [2, 1, 436, 100],
    [2, 2, 436, 200],
    [2, 3, 436, 500],
    [2, 4, 436, 1000],
    [3, 0, 636, 0],
    [3, 1, 636, 100],
    [3, 2, 636, 200],
    [3, 3, 636, 500],
    [3, 4, 636, 1000]
])

shapes_data = [
    [
        [0, 10, 10],
        [0, 10, 20],
        [0, 20, 20],
        [0, 20, 10]
    ],
    [
        [4, 10, 10],
        [4, 10, 20],
        [4, 20, 20],
        [4, 20, 10]
    ]
]

with napari.gui_qt():
    viewer = napari.view_image(test_im, name='image')
    viewer.add_tracks(tracks_data, name='tracks')
    viewer.add_points(tracks_data[:, 1:], name='points')
    viewer.add_shapes(shapes_data, name='shapes')
```
</details>

## Expected behavior

The values for the 0th dimension in `layer._extent_data` should be the same for the image

## Environment
```
napari: 0.3.8.dev14+g504ef2f
Platform: Linux-5.4.0-48-generic-x86_64-with-debian-bullseye-sid
Python: 3.7.7 (default, May 7 2020, 21:25:33) [GCC 7.3.0]
Qt: 5.14.2
PyQt5: 5.14.2
NumPy: 1.19.1
SciPy: 1.5.2
Dask: 2.24.0
VisPy: 0.6.4

GL version: 4.6.0 NVIDIA 450.66
MAX_TEXTURE_SIZE: 32768

Plugins:
- napari_plugin_engine: 0.1.7
- ome_zarr: 0.0.14
- svg: 0.1.4
```
## Additional context

This is related to https://github.com/napari/napari/issues/922

**Comment by sofroniewn** [comment]
> The values for the 0th dimension in layer._extent_data should be the same for the image

Thanks for this write up @kevinyamauchi, I will probably need to ponder this one more, and am curious what @jni thinks, but am I write in thinking that this can be fixed by dropping all the `+1` but making sure that any rounding happening on the sliders always makes sure to include the last slice for dims?

**Comment by jni** [comment]
I'm not sure about the fix, but the reason that it's different for images vs points et al is that images use 0-based half-open indexing, so we wanted to be able to read the dims extent from the layer extent. There are no such concerns with vector quantities, but then the dims slicing needs to know that the extent of a points layer and the extent of an image layer might mean different things. And of course this all becomes a bit hairy once you add a scale to the image.

I guess it might make sense to adopt a convention that extent is *not* half-open, even for images.

**Comment by sofroniewn** [comment]
> that it's different for images vs points et al is that images use 0-based half-open indexing, so we wanted to be able to read the dims extent from the layer extent

Ah yes, good point, especially now with #1660 merged that makes sense to me

**Comment by kevinyamauchi** [comment]
Thanks for the quick responses, @sofroniewn and @jni!

> the reason that it's different for images vs points et al is that images use 0-based half-open indexing, so we wanted to be able to read the dims extent from the layer extent.

Got it - the rationale makes sense to me.

> I guess it might make sense to adopt a convention that extent is not half-open, even for images.

At the moment, I'm not sure if I have a strong preference if the extents should all be half-open intervals or all be closed intervals, but I agree that they should all be the same (i.e., don't do special casing on the dims).

It looks like to make the change we would have to:

- Update the layers so the extents are all the closed interval. I think this would only be the `Image` and `Labels` layer.
- Update the `Dims.nsteps` calculation.
- Update the docs

Am I missing anything? Does that seem right?

**Comment by sofroniewn** [comment]
> Am I missing anything? Does that seem right?

I think that is right, it looks like we just need to add a `-1` to the image shape here https://github.com/napari/napari/blob/master/napari/layers/image/image.py#L332, right? (labels inherits)

I think that will be what we need.

Consider:

```python
with napari.gui_qt():
    viewer = napari.Viewer()
    viewer.add_image(np.random.random((11, 11)), colormap='red', opacity=.5, name='half')
    viewer.add_points([[0, 0], [10, 10]], size=0.5, opacity=.5)
```

<img width="1200" alt="Screen Shot 2020-10-05 at 1 06 32 PM" src="https://user-images.githubusercontent.com/6531703/95127013-2e234f00-070c-11eb-9972-073a55b6df29.png">

Both of these layers should have an extent of `0` to `10`, which means doing a `-1` on the shape for the `Image` layer. Looks like we already have a `+1` on `nsteps`, will we need another one? https://github.com/napari/napari/blob/master/napari/components/dims.py#L107

**Comment by sofroniewn** [comment]
@kevinyamauchi I have created a PR to fix this in #1696
